### PR TITLE
use custom sms provider #385

### PR DESCRIPTION
**Pull request by novanda1**
## What kind of change does this PR introduce?

feature

## What is the current behavior?

[Using custom (Unlimited) SMS Provider for sending OTP](https://github.com/supabase/gotrue/issues/385)

## What is the new behavior?

There is a new optional `boolean` param named "send_otp" that controls whether the API sends the OTP code to the SMS Provider or not, since the default value is `true` it will not break the current behavior.

### Current Behavior
```
curl -X POST 'https://tqfkjdmpohkqylveukuhsy.supabase.net/auth/v1/otp' \
-H "apikey: <SERVICE_KEY>" \
-H "Content-Type: application/json" \
-d '{
  "phone": "6212312312312"
}'
{}
```
### With new Param
```
curl -X POST 'https://tqfkjdmpohkqylveukuhsy.supabase.net/auth/v1/otp' \
-H "apikey: <SERVICE_KEY>" \
-H "Content-Type: application/json" \
-d '{
  "phone": "6212312312312", 
  "send_otp": false
}'
{"otp":"414995"}
```

### CODE REVIEW

**Comment by novanda1** [comment]
(edited)

Recently I realized that the current behavior on "/otp" route is didn't require admin credentials, so adding `send_otp` parameter and returning the OTP code in that URL is not safe.

I decided to add a new route inside of "/otp" that is "/otp/phone" and use `requireAdminCredentials` on it, so the behavior of "/otp" is still the same like current behavior, and when we request to "/otp/phone" it need a `SERVICE_ROLE_KEY`


### Current Behavior
```
curl -X POST 'https://tqfkjdmpohkqylveukuhsy.supabase.net/auth/v1/otp' \
-H "Content-Type: application/json" \
-d '{
  "phone": "6212312312312"
}'
{}
```

### New Route
```
curl -X POST 'https://tqfkjdmpohkqylveukuhsy.supabase.net/auth/v1/otp/phone' \
-H "apikey: <SERVICE_KEY>" \
-H "Content-Type: application/json" \
-d '{
  "phone": "6212312312312", }'
{"otp":"414995"}
```

Honestly, I'm not sure about the route name, and request method.

Please let me know if you have feedback.

**Review by kangmingtay — CHANGES_REQUESTED** [review]
hey @novanda1, thanks for making this PR, you also need to handle the case where an OTP is sent out when a user updates their phone number using `PUT /user`. 

Currently, that might be slightly complicated to handle because `PUT /user` is called with the authenticated user's JWT. However, we'll want to protect this `generate_otp` endpoint behind a service role JWT so it might be better to handle this case on the `PUT /admin/user` endpoint instead.

If you're interested, you can drop your thoughts around the design and the implementation first and we can review it before you get started.

File: api/otp.go [hunk 1]
```diff
@@ -135,3 +141,84 @@ func (a *API) shouldCreateUser(r *http.Request, params *OtpParams) bool {
 	}
 	return true
 }
+
+// GetPhoneOtp save and return the OTP code
+func (a *API) GetPhoneOtp(w http.ResponseWriter, r *http.Request) error {
```
Comment (kangmingtay) [inline api/otp.go:146]: Can we find a way to reuse the `Otp` method above? That way, we won't have to maintain 2 separate functions with almost identical functionality.

**Comment by kangmingtay** [comment]
Hi @novanda1, on second thought, it will be much cleaner for us to modify the `/admin/generate_link` endpoint to one that takes in a `phone` field and `type` field. This would allow users to generate a link or otp for email or phone confirmations. The endpoint name will probably have to be renamed and we'll be discussing this internally to figure out what makes sense.

**Pull request closed without merging**